File: Controls/AngleCalibrationTraceControl.vb
Imports System.ComponentModel
Imports System.Windows.Media
Imports WpfApp1.Models
Imports WpfApp1.Services

Namespace Controls
    Public Class AngleCalibrationTraceControl
        Inherits FrameworkElement

        Private Const ViewPadding As Double = 26.0
        Private Const MinimumScale As Double = 0.08
        Private Const MaximumScale As Double = 1.8
        Private Const HistoryStrokeOpacity As Double = 0.22
        Private Const CurrentStrokeOpacity As Double = 0.55
        Private Const StrokeThickness As Double = 1.65

        Public Shared ReadOnly RenderFrameProperty As DependencyProperty =
            DependencyProperty.Register("RenderFrame",
                                        GetType(AngleCalibrationRenderFrame),
                                        GetType(AngleCalibrationTraceControl),
                                        New FrameworkPropertyMetadata(Nothing,
                                                                      FrameworkPropertyMetadataOptions.AffectsRender,
                                                                      AddressOf OnRenderFrameChanged))

        Private _historyPen As Pen
        Private _currentPen As Pen
        Private _targetCenterX As Double
        Private _targetCenterY As Double
        Private _targetScale As Double = 1.0
        Private _currentCenterX As Double
        Private _currentCenterY As Double
        Private _currentScale As Double = 1.0
        Private _hasViewState As Boolean
        Private _pendingAnimationFrames As Integer
        Private _isRenderingSubscribed As Boolean
        Private _isThemeSubscribed As Boolean

        Public Sub New()
            SnapsToDevicePixels = True
            UseLayoutRounding = True
            ApplyThemeResources()
            AddHandler Loaded, AddressOf OnLoaded
            AddHandler Unloaded, AddressOf OnUnloaded
        End Sub

        Public Property RenderFrame As AngleCalibrationRenderFrame
            Get
                Return CType(GetValue(RenderFrameProperty), AngleCalibrationRenderFrame)
            End Get
            Set(value As AngleCalibrationRenderFrame)
                SetValue(RenderFrameProperty, value)
            End Set
        End Property

        Private Shared Sub OnRenderFrameChanged(d As DependencyObject, e As DependencyPropertyChangedEventArgs)
            CType(d, AngleCalibrationTraceControl).HandleRenderFrameChanged()
        End Sub

        Private Sub HandleRenderFrameChanged()
            RecalculateTargetView()

            If Not _hasViewState Then
                _currentCenterX = _targetCenterX
                _currentCenterY = _targetCenterY
                _currentScale = _targetScale
                _hasViewState = True
            End If

            _pendingAnimationFrames = 10
            UpdateRenderingSubscription()
            InvalidateVisual()
        End Sub

        Private Sub OnLoaded(sender As Object, e As RoutedEventArgs)
            If Not _isThemeSubscribed Then
                AddHandler ThemeManager.Instance.ThemeChanged, AddressOf OnThemeChanged
                _isThemeSubscribed = True
            End If

            ApplyThemeResources()
            RecalculateTargetView()
            UpdateRenderingSubscription()
        End Sub

        Private Sub OnUnloaded(sender As Object, e As RoutedEventArgs)
            If _isThemeSubscribed Then
                RemoveHandler ThemeManager.Instance.ThemeChanged, AddressOf OnThemeChanged
                _isThemeSubscribed = False
            End If

            UpdateRenderingSubscription(forceDetach:=True)
        End Sub

        Private Sub OnThemeChanged(sender As Object, e As EventArgs)
            ApplyThemeResources()
            InvalidateVisual()
        End Sub

        Private Sub UpdateRenderingSubscription(Optional forceDetach As Boolean = False)
            Dim shouldRender = Not forceDetach AndAlso IsLoaded AndAlso (_pendingAnimationFrames > 0 OrElse (RenderFrame IsNot Nothing AndAlso RenderFrame.IsLocked))

            If Not shouldRender Then
                If _isRenderingSubscribed Then
                    RemoveHandler CompositionTarget.Rendering, AddressOf OnRendering
                    _isRenderingSubscribed = False
                End If

                Return
            End If

            If Not _isRenderingSubscribed Then
                AddHandler CompositionTarget.Rendering, AddressOf OnRendering
                _isRenderingSubscribed = True
            End If
        End Sub

        Private Sub OnRendering(sender As Object, e As EventArgs)
            If Not _hasViewState Then
                UpdateRenderingSubscription(forceDetach:=True)
                Return
            End If

            Dim alpha = If(RenderFrame IsNot Nothing AndAlso RenderFrame.IsLocked, 0.18, 0.24)
            _currentCenterX += (_targetCenterX - _currentCenterX) * alpha
            _currentCenterY += (_targetCenterY - _currentCenterY) * alpha
            _currentScale += (_targetScale - _currentScale) * alpha

            If Math.Abs(_targetCenterX - _currentCenterX) < 0.15 AndAlso
               Math.Abs(_targetCenterY - _currentCenterY) < 0.15 AndAlso
               Math.Abs(_targetScale - _currentScale) < 0.002 Then
                _currentCenterX = _targetCenterX
                _currentCenterY = _targetCenterY
                _currentScale = _targetScale
                _pendingAnimationFrames = Math.Max(0, _pendingAnimationFrames - 1)
            Else
                _pendingAnimationFrames = Math.Max(_pendingAnimationFrames, 2)
            End If

            InvalidateVisual()
            UpdateRenderingSubscription()
        End Sub

        Protected Overrides Sub OnRender(drawingContext As DrawingContext)
            MyBase.OnRender(drawingContext)

            Dim renderFrameValue = RenderFrame
            If renderFrameValue Is Nothing OrElse ActualWidth <= 0 OrElse ActualHeight <= 0 Then
                Return
            End If

            If Not _hasViewState Then
                RecalculateTargetView()
                _currentCenterX = _targetCenterX
                _currentCenterY = _targetCenterY
                _currentScale = _targetScale
                _hasViewState = True
            End If

            drawingContext.PushClip(New RectangleGeometry(New Rect(0.0, 0.0, ActualWidth, ActualHeight)))

            For Each stroke In renderFrameValue.TraceStrokes
                DrawStroke(drawingContext, stroke)
            Next

            drawingContext.Pop()
        End Sub

        Private Sub DrawStroke(drawingContext As DrawingContext, stroke As AngleCalibrationTraceStroke)
            If stroke Is Nothing OrElse stroke.Points Is Nothing OrElse stroke.Points.Count < 2 Then
                Return
            End If

            Dim geometry As New StreamGeometry()
            Using context = geometry.Open()
                Dim firstPoint = ToScreenPoint(stroke.Points(0))
                context.BeginFigure(firstPoint, False, False)

                For pointIndex = 1 To stroke.Points.Count - 1
                    context.LineTo(ToScreenPoint(stroke.Points(pointIndex)), True, True)
                Next
            End Using

            geometry.Freeze()

            Dim pen = If(stroke.IsCurrent, _currentPen, _historyPen)
            drawingContext.PushOpacity(If(stroke.IsCurrent, CurrentStrokeOpacity, HistoryStrokeOpacity))
            drawingContext.DrawGeometry(Nothing, pen, geometry)
            drawingContext.Pop()
        End Sub

        Private Function ToScreenPoint(point As AngleCalibrationTracePoint) As Point
            Dim x = ((point.X - _currentCenterX) * _currentScale) + (ActualWidth / 2.0)
            Dim y = ((point.Y - _currentCenterY) * _currentScale) + (ActualHeight / 2.0)
            Return New Point(x, y)
        End Function

        Private Sub RecalculateTargetView()
            Dim minX = -200.0
            Dim maxX = 200.0
            Dim minY = -80.0
            Dim maxY = 80.0
            Dim hasPoint = False

            Dim renderFrameValue = RenderFrame
            If renderFrameValue IsNot Nothing AndAlso renderFrameValue.TraceStrokes IsNot Nothing Then
                For Each stroke In renderFrameValue.TraceStrokes
                    If stroke Is Nothing OrElse stroke.Points Is Nothing Then
                        Continue For
                    End If

                    For Each point In stroke.Points
                        If Not hasPoint Then
                            minX = point.X
                            maxX = point.X
                            minY = point.Y
                            maxY = point.Y
                            hasPoint = True
                        Else
                            minX = Math.Min(minX, point.X)
                            maxX = Math.Max(maxX, point.X)
                            minY = Math.Min(minY, point.Y)
                            maxY = Math.Max(maxY, point.Y)
                        End If
                    Next
                Next
            End If

            Dim spanX = Math.Max(1.0, maxX - minX)
            Dim spanY = Math.Max(1.0, maxY - minY)
            Dim availableWidth = Math.Max(32.0, ActualWidth - (ViewPadding * 2.0))
            Dim availableHeight = Math.Max(32.0, ActualHeight - (ViewPadding * 2.0))
            Dim fitScale = Math.Min(availableWidth / spanX, availableHeight / spanY)

            _targetScale = Math.Max(MinimumScale, Math.Min(MaximumScale, fitScale))
            _targetCenterX = (minX + maxX) / 2.0
            _targetCenterY = (minY + maxY) / 2.0
        End Sub

        Private Sub ApplyThemeResources()
            _historyPen = CreatePen(ResolveColor("TextMutedColor", Color.FromRgb(&H98, &H98, &H98)))
            _currentPen = CreatePen(ResolveColor("TextStrongColor", Color.FromRgb(&HC8, &HC8, &HC8)))
        End Sub

        Private Shared Function CreatePen(color As Color) As Pen
            Dim pen As New Pen(New SolidColorBrush(color), StrokeThickness) With {
                .StartLineCap = PenLineCap.Round,
                .EndLineCap = PenLineCap.Round,
                .LineJoin = PenLineJoin.Round
            }

            If pen.Brush.CanFreeze Then
                pen.Brush.Freeze()
            End If

            pen.Freeze()
            Return pen
        End Function

        Private Shared Function ResolveColor(resourceKey As String, fallback As Color) As Color
            Dim resource As Object = Nothing
            If Application.Current IsNot Nothing Then
                resource = Application.Current.TryFindResource(resourceKey)
            End If

            If TypeOf resource Is Color Then
                Return CType(resource, Color)
            End If

            Dim brush = TryCast(resource, SolidColorBrush)
            If brush IsNot Nothing Then
                Return brush.Color
            End If

            Return fallback
        End Function
    End Class
End Namespace

File: src/ClickSyncMouseTester/Controls/AngleCalibrationTraceControl.cs
using ClickSyncMouseTester.Models;
using ClickSyncMouseTester.Services;
using System;
using System.Collections.Generic;
using System.Windows;
using System.Windows.Media;

namespace ClickSyncMouseTester.Controls;

public class AngleCalibrationTraceControl : FrameworkElement
{
    private sealed class CachedTraceStrokeGeometry
    {
        public StreamGeometry Geometry { get; }

        public Rect DataBounds { get; }

        public bool HasDataBounds { get; }

        public CachedTraceStrokeGeometry(StreamGeometry geometry, Rect dataBounds, bool hasDataBounds)
        {
            Geometry = geometry;
            DataBounds = dataBounds;
            HasDataBounds = hasDataBounds;
        }
    }

    private sealed class TraceGeometryEntry
    {
        public StreamGeometry Geometry { get; }

        public bool IsCurrent { get; }

        public TraceGeometryEntry(StreamGeometry geometry, bool isCurrent)
        {
            Geometry = geometry;
            IsCurrent = isCurrent;
        }
    }

    private sealed class TraceGeometryFrame
    {
        public IReadOnlyList<TraceGeometryEntry> Strokes { get; }

        public Rect DataBounds { get; }

        public bool HasDataBounds { get; }

        public TraceGeometryFrame(IReadOnlyList<TraceGeometryEntry> strokes, Rect dataBounds, bool hasDataBounds)
        {
            Strokes = strokes;
            DataBounds = dataBounds;
            HasDataBounds = hasDataBounds;
        }
    }

    private const double ViewPadding = 26.0;

    private const double MinimumScale = 0.08;

    private const double MaximumScale = 1.8;

    private const double HistoryStrokeOpacity = 0.22;

    private const double CurrentStrokeOpacity = 0.55;

    private const double StrokeThickness = 1.65;

    private const double ViewCenterTolerance = 0.15;

    private const double ViewScaleTolerance = 0.002;

    public static readonly DependencyProperty RenderFrameProperty = DependencyProperty.Register("RenderFrame", typeof(AngleCalibrationRenderFrame), typeof(AngleCalibrationTraceControl), new FrameworkPropertyMetadata(null, FrameworkPropertyMetadataOptions.AffectsRender, new PropertyChangedCallback(OnRenderFrameChanged)));

    private readonly Dictionary<AngleCalibrationTraceStroke, CachedTraceStrokeGeometry> _strokeGeometryCache;

    private Pen _historyPen;

    private Pen _currentPen;

    private Color _historyPenColor;

    private Color _currentPenColor;

    private double _penThicknessKey;

    private AngleCalibrationRenderFrame _cachedGeometryFrameSource;

    private TraceGeometryFrame _cachedGeometryFrame;

    private RectangleGeometry _clipGeometry;

    private double _clipWidth;

    private double _clipHeight;

    private double _targetCenterX;

    private double _targetCenterY;

    private double _targetScale;

    private double _currentCenterX;

    private double _currentCenterY;

    private double _currentScale;

    private bool _hasViewState;

    private bool _isRenderingSubscribed;

    private bool _isThemeSubscribed;

    public AngleCalibrationRenderFrame RenderFrame
    {
        get
        {
            return (AngleCalibrationRenderFrame)GetValue(RenderFrameProperty);
        }
        set
        {
            SetValue(RenderFrameProperty, value);
        }
    }

    public AngleCalibrationTraceControl()
    {
        _strokeGeometryCache = new Dictionary<AngleCalibrationTraceStroke, CachedTraceStrokeGeometry>();
        _targetScale = 1.0;
        _currentScale = 1.0;
        _penThicknessKey = double.NaN;
        base.SnapsToDevicePixels = true;
        base.UseLayoutRounding = true;
        ApplyThemeResources();
        base.Loaded += OnLoaded;
        base.Unloaded += OnUnloaded;
    }

    private static void OnRenderFrameChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
    {
        ((AngleCalibrationTraceControl)d).HandleRenderFrameChanged();
    }

    private void HandleRenderFrameChanged()
    {
        RecalculateTargetView();
        if (!_hasViewState)
        {
            _currentCenterX = _targetCenterX;
            _currentCenterY = _targetCenterY;
            _currentScale = _targetScale;
            _hasViewState = true;
        }
        else
        {
            SnapCurrentViewIfClose();
        }
        UpdateRenderingSubscription();
        InvalidateVisual();
    }

    private void OnLoaded(object sender, RoutedEventArgs e)
    {
        if (!_isThemeSubscribed)
        {
            ThemeManager.Instance.ThemeChanged += OnThemeChanged;
            _isThemeSubscribed = true;
        }
        ApplyThemeResources();
        RecalculateTargetView();
        UpdateRenderingSubscription();
    }

    private void OnUnloaded(object sender, RoutedEventArgs e)
    {
        if (_isThemeSubscribed)
        {
            ThemeManager.Instance.ThemeChanged -= OnThemeChanged;
            _isThemeSubscribed = false;
        }
        UpdateRenderingSubscription(forceDetach: true);
    }

    private void OnThemeChanged(object sender, EventArgs e)
    {
        ApplyThemeResources();
        InvalidateVisual();
    }

    private void UpdateRenderingSubscription(bool forceDetach = false)
    {
        if (forceDetach || !base.IsLoaded || !IsViewAnimationActive())
        {
            if (_isRenderingSubscribed)
            {
                CompositionTarget.Rendering -= OnRendering;
                _isRenderingSubscribed = false;
            }
        }
        else if (!_isRenderingSubscribed)
        {
            CompositionTarget.Rendering += OnRendering;
            _isRenderingSubscribed = true;
        }
    }

    private void OnRendering(object sender, EventArgs e)
    {
        if (!_hasViewState)
        {
            UpdateRenderingSubscription(forceDetach: true);
            return;
        }
        double interpolationFactor = RenderFrame != null && RenderFrame.IsLocked ? 0.18 : 0.24;
        _currentCenterX += (_targetCenterX - _currentCenterX) * interpolationFactor;
        _currentCenterY += (_targetCenterY - _currentCenterY) * interpolationFactor;
        _currentScale += (_targetScale - _currentScale) * interpolationFactor;
        SnapCurrentViewIfClose();
        InvalidateVisual();
        UpdateRenderingSubscription();
    }

    protected override void OnRenderSizeChanged(SizeChangedInfo sizeInfo)
    {
        base.OnRenderSizeChanged(sizeInfo);
        _clipGeometry = null;
        RecalculateTargetView();
        if (!_hasViewState && RenderFrame != null)
        {
            _currentCenterX = _targetCenterX;
            _currentCenterY = _targetCenterY;
            _currentScale = _targetScale;
            _hasViewState = true;
        }
        else
        {
            SnapCurrentViewIfClose();
        }
        UpdateRenderingSubscription();
        InvalidateVisual();
    }

    protected override void OnRender(DrawingContext drawingContext)
    {
        base.OnRender(drawingContext);
        AngleCalibrationRenderFrame renderFrame = RenderFrame;
        if (renderFrame == null || base.ActualWidth <= 0.0 || base.ActualHeight <= 0.0)
        {
            return;
        }
        if (!_hasViewState)
        {
            RecalculateTargetView();
            _currentCenterX = _targetCenterX;
            _currentCenterY = _targetCenterY;
            _currentScale = _targetScale;
            _hasViewState = true;
        }
        TraceGeometryFrame geometryFrame = EnsureGeometryFrame(renderFrame);
        if (geometryFrame == null || geometryFrame.Strokes.Count == 0)
        {
            return;
        }
        EnsurePensForScale(_currentScale);
        double offsetX = base.ActualWidth / 2.0 - _currentCenterX * _currentScale;
        double offsetY = base.ActualHeight / 2.0 - _currentCenterY * _currentScale;
        drawingContext.PushClip(GetClipGeometry());
        drawingContext.PushTransform(new MatrixTransform(new Matrix(_currentScale, 0.0, 0.0, _currentScale, offsetX, offsetY)));
        foreach (TraceGeometryEntry traceStroke in geometryFrame.Strokes)
        {
            DrawStroke(drawingContext, traceStroke);
        }
        drawingContext.Pop();
        drawingContext.Pop();
    }

    private void DrawStroke(DrawingContext drawingContext, TraceGeometryEntry stroke)
    {
        if (stroke == null || stroke.Geometry == null)
        {
            return;
        }

        Pen pen = stroke.IsCurrent ? _currentPen : _historyPen;
        drawingContext.PushOpacity(stroke.IsCurrent ? CurrentStrokeOpacity : HistoryStrokeOpacity);
        drawingContext.DrawGeometry(null, pen, stroke.Geometry);
        drawingContext.Pop();
    }

    private TraceGeometryFrame EnsureGeometryFrame(AngleCalibrationRenderFrame renderFrame)
    {
        if (renderFrame == null)
        {
            _cachedGeometryFrameSource = null;
            _cachedGeometryFrame = null;
            _strokeGeometryCache.Clear();
            return null;
        }
        if (ReferenceEquals(_cachedGeometryFrameSource, renderFrame))
        {
            return _cachedGeometryFrame;
        }

        TraceGeometryFrame geometryFrame = CreateGeometryFrame(renderFrame);
        _cachedGeometryFrameSource = renderFrame;
        _cachedGeometryFrame = geometryFrame;
        return geometryFrame;
    }

    private TraceGeometryFrame CreateGeometryFrame(AngleCalibrationRenderFrame renderFrame)
    {
        List<TraceGeometryEntry> strokeEntries = new List<TraceGeometryEntry>();
        HashSet<AngleCalibrationTraceStroke> liveStrokes = new HashSet<AngleCalibrationTraceStroke>();
        Rect dataBounds = Rect.Empty;
        bool hasDataBounds = false;
        if (renderFrame.TraceStrokes != null)
        {
            foreach (AngleCalibrationTraceStroke traceStroke in renderFrame.TraceStrokes)
            {
                if (traceStroke == null)
                {
                    continue;
                }
                liveStrokes.Add(traceStroke);
                CachedTraceStrokeGeometry cachedStrokeGeometry = GetOrCreateStrokeGeometry(traceStroke);
                if (cachedStrokeGeometry == null)
                {
                    continue;
                }
                if (cachedStrokeGeometry.HasDataBounds)
                {
                    IncludeBounds(ref dataBounds, ref hasDataBounds, cachedStrokeGeometry.DataBounds);
                }
                if (cachedStrokeGeometry.Geometry != null)
                {
                    strokeEntries.Add(new TraceGeometryEntry(cachedStrokeGeometry.Geometry, traceStroke.IsCurrent));
                }
            }
        }
        RemoveStaleStrokeGeometry(liveStrokes);
        return new TraceGeometryFrame(strokeEntries, dataBounds, hasDataBounds);
    }

    private CachedTraceStrokeGeometry GetOrCreateStrokeGeometry(AngleCalibrationTraceStroke traceStroke)
    {
        if (_strokeGeometryCache.TryGetValue(traceStroke, out CachedTraceStrokeGeometry cachedGeometry))
        {
            return cachedGeometry;
        }
        cachedGeometry = CreateStrokeGeometry(traceStroke);
        _strokeGeometryCache[traceStroke] = cachedGeometry;
        return cachedGeometry;
    }

    private static CachedTraceStrokeGeometry CreateStrokeGeometry(AngleCalibrationTraceStroke traceStroke)
    {
        if (traceStroke == null || traceStroke.Points == null || traceStroke.Points.Count == 0)
        {
            return null;
        }

        Rect dataBounds = Rect.Empty;
        bool hasDataBounds = false;
        foreach (AngleCalibrationTracePoint point in traceStroke.Points)
        {
            IncludePoint(ref dataBounds, ref hasDataBounds, point);
        }

        StreamGeometry strokeGeometry = null;
        if (traceStroke.Points.Count >= 2)
        {
            strokeGeometry = new StreamGeometry();
            using (StreamGeometryContext geometryContext = strokeGeometry.Open())
            {
                AngleCalibrationTracePoint startPoint = traceStroke.Points[0];
                geometryContext.BeginFigure(new Point(startPoint.X, startPoint.Y), isFilled: false, isClosed: false);
                for (int pointIndex = 1; pointIndex < traceStroke.Points.Count; pointIndex++)
                {
                    AngleCalibrationTracePoint point = traceStroke.Points[pointIndex];
                    geometryContext.LineTo(new Point(point.X, point.Y), isStroked: true, isSmoothJoin: true);
                }
            }
            if (strokeGeometry.CanFreeze)
            {
                strokeGeometry.Freeze();
            }
        }

        return new CachedTraceStrokeGeometry(strokeGeometry, dataBounds, hasDataBounds);
    }

    private void RemoveStaleStrokeGeometry(HashSet<AngleCalibrationTraceStroke> liveStrokes)
    {
        List<AngleCalibrationTraceStroke> staleStrokes = null;
        foreach (AngleCalibrationTraceStroke cachedStroke in _strokeGeometryCache.Keys)
        {
            if (!liveStrokes.Contains(cachedStroke))
            {
                if (staleStrokes == null)
                {
                    staleStrokes = new List<AngleCalibrationTraceStroke>();
                }
                staleStrokes.Add(cachedStroke);
            }
        }
        if (staleStrokes == null)
        {
            return;
        }
        foreach (AngleCalibrationTraceStroke staleStroke in staleStrokes)
        {
            _strokeGeometryCache.Remove(staleStroke);
        }
    }

    private static void IncludePoint(ref Rect dataBounds, ref bool hasDataBounds, AngleCalibrationTracePoint point)
    {
        if (point == null)
        {
            return;
        }
        IncludePoint(ref dataBounds, ref hasDataBounds, point.X, point.Y);
    }

    private static void IncludePoint(ref Rect dataBounds, ref bool hasDataBounds, double x, double y)
    {
        if (!hasDataBounds)
        {
            dataBounds = new Rect(new Point(x, y), new Size(0.0, 0.0));
            hasDataBounds = true;
            return;
        }
        dataBounds.Union(new Point(x, y));
    }

    private static void IncludeBounds(ref Rect dataBounds, ref bool hasDataBounds, Rect bounds)
    {
        if (bounds.IsEmpty)
        {
            return;
        }
        if (!hasDataBounds)
        {
            dataBounds = bounds;
            hasDataBounds = true;
            return;
        }
        dataBounds.Union(bounds);
    }

    private void RecalculateTargetView()
    {
        double minX = -200.0;
        double maxX = 200.0;
        double minY = -80.0;
        double maxY = 80.0;
        TraceGeometryFrame geometryFrame = EnsureGeometryFrame(RenderFrame);
        if (geometryFrame != null && geometryFrame.HasDataBounds)
        {
            minX = geometryFrame.DataBounds.Left;
            maxX = geometryFrame.DataBounds.Right;
            minY = geometryFrame.DataBounds.Top;
            maxY = geometryFrame.DataBounds.Bottom;
        }

        double contentWidth = Math.Max(1.0, maxX - minX);
        double contentHeight = Math.Max(1.0, maxY - minY);
        double availableWidth = Math.Max(32.0, base.ActualWidth - ViewPadding * 2.0);
        double availableHeight = Math.Max(32.0, base.ActualHeight - ViewPadding * 2.0);
        double scaleToFit = Math.Min(availableWidth / contentWidth, availableHeight / contentHeight);
        _targetScale = Math.Max(MinimumScale, Math.Min(MaximumScale, scaleToFit));
        _targetCenterX = (minX + maxX) / 2.0;
        _targetCenterY = (minY + maxY) / 2.0;
    }

    private bool IsViewAnimationActive()
    {
        return _hasViewState && !IsCurrentViewCloseToTarget();
    }

    private bool IsCurrentViewCloseToTarget()
    {
        return Math.Abs(_targetCenterX - _currentCenterX) < ViewCenterTolerance
            && Math.Abs(_targetCenterY - _currentCenterY) < ViewCenterTolerance
            && Math.Abs(_targetScale - _currentScale) < ViewScaleTolerance;
    }

    private void SnapCurrentViewIfClose()
    {
        if (!IsCurrentViewCloseToTarget())
        {
            return;
        }
        _currentCenterX = _targetCenterX;
        _currentCenterY = _targetCenterY;
        _currentScale = _targetScale;
    }

    private RectangleGeometry GetClipGeometry()
    {
        double actualWidth = base.ActualWidth;
        double actualHeight = base.ActualHeight;
        if (_clipGeometry != null && Math.Abs(_clipWidth - actualWidth) < 0.001 && Math.Abs(_clipHeight - actualHeight) < 0.001)
        {
            return _clipGeometry;
        }
        _clipWidth = actualWidth;
        _clipHeight = actualHeight;
        _clipGeometry = new RectangleGeometry(new Rect(0.0, 0.0, actualWidth, actualHeight));
        if (_clipGeometry.CanFreeze)
        {
            _clipGeometry.Freeze();
        }
        return _clipGeometry;
    }

    private void ApplyThemeResources()
    {
        _historyPenColor = ResolveColor("TextMutedColor", Color.FromRgb(152, 152, 152));
        _currentPenColor = ResolveColor("TextStrongColor", Color.FromRgb(200, 200, 200));
        _penThicknessKey = double.NaN;
        EnsurePensForScale(_currentScale);
    }

    private void EnsurePensForScale(double scale)
    {
        double safeScale = Math.Max(0.001, Math.Abs(scale));
        double penThickness = StrokeThickness / safeScale;
        double penThicknessKey = Math.Round(penThickness * 64.0) / 64.0;
        if (_historyPen != null && _currentPen != null && Math.Abs(_penThicknessKey - penThicknessKey) < 0.0001)
        {
            return;
        }

        _penThicknessKey = penThicknessKey;
        _historyPen = CreatePen(_historyPenColor, penThicknessKey);
        _currentPen = CreatePen(_currentPenColor, penThicknessKey);
    }

    private static Pen CreatePen(Color color, double thickness)
    {
        Pen pen = new Pen(new SolidColorBrush(color), thickness)
        {
            StartLineCap = PenLineCap.Round,
            EndLineCap = PenLineCap.Round,
            LineJoin = PenLineJoin.Round
        };
        if (pen.Brush.CanFreeze)
        {
            pen.Brush.Freeze();
        }
        if (pen.CanFreeze)
        {
            pen.Freeze();
        }
        return pen;
    }

    private static Color ResolveColor(string resourceKey, Color fallback)
    {
        object resource = null;
        if (System.Windows.Application.Current != null)
        {
            resource = System.Windows.Application.Current.TryFindResource(resourceKey);
        }
        if (resource is Color colorValue)
        {
            return colorValue;
        }
        if (!(resource is SolidColorBrush { Color: var color }))
        {
            return fallback;
        }
        return color;
    }
}
